pkg/kubernetes/nulecule.js

```diff
@@ -0,0 +1,182 @@
+            var process = cockpit.spawn(["/usr/bin/atomicapp", "-d", "install", "--destination", tmp_dir, image]);
+
+            console.log("installing image: " + image + " in folder "+tmp_dir)
+            deferred.notify(_("Installing Application..."));
+
+            process.always(function() {
+                    console.log("....always.....");
+                    //window.clearInterval(timer);
+                })
+                .stream(function(text) {
+                    buffer += text;
+                    console.log("buf = "+buffer);
+                    deferred.notify(buffer);                   
+                })
+                .done(function(output) {
+                    console.log("....done.....");
+                    deferred.resolve();
+                })
+                .fail(function(ex) {
+                    console.log("....fail.....");
+                    var message;
+                    if (ex.problem === "cancelled") {
+                        deferred.reject(ex);
+                        return;
+                    }
+
+                    if (!message) {
+                        message = _("Image failed to Install");
+                        console.warn(ex.message);
+                    }
+                    var error = new Error(message);
+                    deferred.reject(error);
+                });
+
+            promise = deferred.promise();
+            promise.cancel = function cancel() {
+                console.log("....cancelled.....");
+                //window.clearInterval(timer);
+                process.close("cancelled");
+            };
+
+            return promise;
+        };
+    }
+
+    
+    /*
+     * Returns a new instance of Constructor for each
+     * key passed into the returned function. Multiple
+     * callers for the same key will get the same instance.
+     *
+     * Overrides .close() on the instances, to close when
+     * all callers have closed.
+     *
+     * Instances must accept zero or one primitive arguments,
+     * and must have zero arguments in their .close() method.
+     */
+     //TODO move it 
+    function singleton(Constructor) {
+        var cached = { };
+
+        return function(key) {
+            var str = key + "";
+
+            var item = cached[str];
+            if (item) {
+                item.refs += 1;
+                return item.obj;
+            }
+
+            item = { refs: 1, obj: new Constructor(key) };
+            var close = item.obj.close;
+            item.obj.close = function close_singleton() {
+                item.refs -= 1;
+                if (item.refs === 0) {
+                    delete cached[str];
+                    if (close)
+                        close.apply(item.obj);
+                }
+            };
+
+            cached[str] = item;
+            return item.obj;
+        };
+    }
+
+    nulecule.nuleculeclient = singleton(NuleculeClient);
```

Comment:
Why is the nulecule client a singleton? Does it need to have this complexity? Are there going to be several parts of cockpit sharing a single NuleculeClient?